src/java.base/share/classes/java/lang/ScopedValue.java

```diff
@@ -698,18 +702,28 @@ private static synchronized int generateKey() {
             x ^= x >>> 12;
             x ^= x << 9;
             x ^= x >>> 23;
-        } while (Cache.primarySlot(x) == Cache.secondarySlot(x));
+        } while (((Cache.primaryIndex(x) ^ Cache.secondaryIndex(x)) & 1) == 0);
         return (nextKey = x);
     }
 
+    // This method exists only so that we can generate keys early in
+    // the boot cycle, before the class j.u.c.ThreadLocalRandom has
+    // been initialized.
```

Comment:
I recommend also adding "or System.getProperty is accessible" here. I think we need to describe in detail how this dependency system works:

1. Cache size depends on System.getProperty (Reason for original failure)
2. Hash depends on ThreadLocalRandom initialization

So we want to convey that in early boot, we can create SV, but we cannot bind/query it. No matter what implementation class model we choose, I think we should document this two-stage initialization process of:

1. Early boot, can create SV, cannot use it
2. 2 dependencies ready, can now bind/query SV.

Reply:
Thanks. Is this better?